Basic setup_keywords integration

nils-werner opened the pull request:
This implementation is inspired by [CFFI's setuptools integration](https://cffi.readthedocs.org/en/latest/cdef.html#distutils-setuptools).

They do not require you to `import cffi` but simply define an additional keyword argument `cffi_modules` for `setup()`. This additional keyword argument does not raise errors when cffi is not installed yet but can be used once cffi is there (as defined in `setup_requires`). Later setuptools will call cffi and have it do whatever it wants with the contents of the argument.

Cython can do the exact same. Below you have the usual `setup.py`, with all the chicken and egg `import` problems and a few `Extensions` that need to be `cythonize`d:

    import setuptools
    from Cython.Build import cythonize
    from distutils.extension import Extension

    extensions = [
        Extension("fib", ["fib.pyx"]),
        Extension("fib2", ["fib2.pyx"]),
    ]

    setuptools.setup(
        name='example',
        version="1.0.0",
        packages=setuptools.find_packages(),
        setup_requires=[
            'cython',
        ],
        ext_modules=cythonize(extensions),
    )

With this kind of setuptools integration the usual

    ext_modules=cythonize(extensions),

can be replaced with

    cython_modules=extensions,

removing the need for `from Cython.Build import cythonize` and solving the import problem:

    import setuptools
    from distutils.extension import Extension

    extensions = [
        Extension("fib", ["fib.pyx"]),
        Extension("fib2", ["fib2.pyx"]),
    ]

    setuptools.setup(
        name='example',
        version="1.0.0",
        setup_requires=['cython',],  # we must later require a Cython version that has this kind of integration
        install_requires=['cython',],
        cython_modules=extensions,
    )

Now I have superficially tested this and it is working fine. However there might be some edgecases which I am unaware of. So please see this as a conversation starter rather than a production ready pull request.

robertwb commented:
Very cool. What happens when Cython is not available? (The current setup allows one to ship the .c files with a dummy cythonize.)

nils-werner commented:
You can define your `setup.py` like

    import setuptools
    from setuptools.extension import Extension

    setuptools.setup(
        name='example',
        version="1.0.0",
        extras_require={'cython': ['cython']},
        ext_modules=[
            Extension("fib", ["fib.c"]),
            Extension("fib2", ["fib2.c"]),
        ],
        cython_modules=[
            Extension("fib", ["fib.pyx"]),
            Extension("fib2", ["fib2.pyx"]),
        ],
    )

When cython does not exist it will only compile the `ext_modules` extensions.

When cython exists it will **compile both `ext_modules` and `cython_modules`**. In this simple example it was no problem but it may be one for larger projects. What is the desired behaviour when you have cython vs when you don't have it?

nils-werner commented:
Depending on how strict you are about having a DRY `setup.py` you may not like the double declaration of the `Extension(... *.pyx)` and their corresponding `Extension(... *.c)` bits.

I personally don't mind repetitive declarations as in most cases they are

 - the most appropriate representation of the values they contain
 - [more readable](https://github.com/nils-werner/statsmodels/blob/setuppy/setup.py) than any [attempts to reduce repetition](https://github.com/statsmodels/statsmodels/blob/master/setup.py)

But I can see that this might be an issue to some and that a declaration like

    import setuptools
    from setuptools.extension import Extension
    
    extensions = [
        Extension("fib", ["fib.c"]),
        Extension("fib2", ["fib2.c"]),
    ]
    
    setuptools.setup(
        name='example',
        version="1.0.0",
        extras_require={'cython': ['cython']},
        ext_modules=extensions,
        cython_modules=extensions,
    )

might seem nicer.

To solve this I however would rather not ask the user to implement some heuristic in their `setup.py` but instead extend `cythonize`. Right now `cythonize` bypasses entirely when you pass it anything but a `*.pyx` or `*.py` file.

If we were to implement some heuristic I would rather change the behavior of `cythonize`:

 - Look for a `foo.pyx` file even when passed `foo.c`
 - If exists: treat Extension as if it was `foo.pyx`
 - Continue as usual

That way you can gracefully and transparently extend the default `Extension` behavior to automatically build using cython, if installed and fall back to plain `Extension`, if not.

All with having zero boilerplate for the user.

nils-werner commented:
Actually, because `setuptools` is designed to allow multiple packages to attach to the same keyword we can go totally crazy and extend `ext_modules` itself. Combined with the `cythonize` heuristics suggested above you can have a `setup.py` like

    import setuptools
    from setuptools.extension import Extension

    setuptools.setup(
        name='example',
        version="1.0.0",
        extras_require={'cython': ['cython']},
        ext_modules=[
            Extension("fib", ["fib.c"]),
            Extension("fib2", ["fib2.c"]),
        ],
    )

that automatically handles both situations (cython installed or not installed) 100% transparently and with absolutely, truly zero boilerplate.

nils-werner commented:
Any thoughts or updates?

robertwb commented:
Thinking about this some more, I actually prefer the 

    ext_modules=cythonize(extensions),

format. In particular, letting it go glob file expansion and extract distutils parameters. If we provided a public-domain cythonize one could distribute with your project that would be idea (provided it could be kept small enough). 

The look-for-corresponding-pyx-file-for-each-c-file approach is quite intriguing though (but, as above, it may miss things). I think this deserves discussion in a wider group; could you pose these ideas to the mailing list?

jakirkham commented:
If you do search for `pyx` when `c` is given, it probably should be extended to the C++ variants IMO.

nils-werner commented:
@robertwb

    cython_modules=extensions,

is just shorthand for

    ext_modules=cythonize(extensions),

So globbing etc. still works (additional keyword arguments don't, but judging from the docs they're only shorthand for `cythonize(Extension(..., keyword_arguments))` anyways).

What exactly do you mean by *extracting distutils parameters*?

And what do you mean by

> If we provided a public-domain cythonize one could distribute with your project that would be idea (provided it could be kept small enough).

@jakirkham: it replaces any extension so C++ stuff will be replaced, too.

jakirkham commented:
Thanks for clarifying.

robertwb commented:
What I mean by "extracting distutils parameters " is that cythonize() extracts external dependencies, compiler flags, and other distutils options form the sources, see https://github.com/cython/cython/tree/master/tests/build Thus just listing the exceptions by merely replacing ".pyx" with ".c" isn't enough.

rocktavious commented:
Has there been any movement on this?  I would really love this feature, at first i started to write this on my own (https://github.com/rocktavious/setuptools.cython - coming at it from the automatic detection of pyx files), and through some spelunking i landed here.  I would gladly help push this through if we have some consensus or direction.  I think the 90% common usecase is to have it work very similar to setuptools' find_packages() method (that also cythonizes), and then provide the necessary hooks/args to configure for the 10% edge cases.

jakirkham commented:
Also, kind of related change in setuptools ( https://bitbucket.org/pypa/setuptools/commits/424966904023 ).

honnibal commented:
> When cython exists it will only build each pyx file if it changed and then compile each Extension once

In case this isn't covered currently: be careful of changes to .pxd files. If a .pxd file is changed, all of the .pyx files will need to be recompiled, because there's no (easy) way to tell what files ultimately depend on the modified header.

hameerabbasi commented:
I realize this PR is old but I'd love to have this in Cython. Is there an interest in making this a reality?

scoder commented:
I can see that this solves at least one of the problems that users have with Cython integration: the fact that distutils/setuptools should decide when to call `cythonize()` instead of always having it executed, even on non-build runs. So this solves an actual problem in an IMHO reasonable way. It does not tackle the problem of not having Cython installed, but that's a bigger problem alltogether.

What do others think about allowing a dict as argument, and passing that as keyword arguments into `cythonize()`? That would enable more of the existing build functionality.

nils-werner commented:
I've rebased the branch to `master`.

jakirkham commented on Cython/Setuptools.py:
```diff
@@ -0,0 +1,36 @@
+    pass
+
+
+class BuildPyx(Command):
```
Should we be subclassing `build_ext`?

scoder replied:
I think not, because this command should run *before* `build_ext`.

scoder commented on Cython/Build/Dependencies.py:
```diff
@@ -881,7 +881,7 @@ class Extension_setuptools(Extension): pass
 # This is the user-exposed entry point.
 def cythonize(module_list, exclude=None, nthreads=0, aliases=None, quiet=False, force=False, language=None,
-              exclude_failures=False, **options):
+              exclude_failures=False, replace_extension=False, **options):
```
`replace_extension` isn't the ideal name because it reads in this context like we're replacing the extension (module), not the (file) extension. Maybe `auto_use_pyx`? And that should obviously be documented in the compilation docs.

nils-werner replied:
I've changed it to `auto_use_pyx` and documented it in the docstring. Anywhere else that I missed?

scoder commented:
Sorry for breaking `master` while you were rebasing. Could you please rebase again, and also provide some tests that show that this works and how it can be used?

nils-werner commented:
I've rebased it again and added a few very simple tests.

nils-werner commented on Cython/Setuptools.py:
```diff
@@ -0,0 +1,36 @@
+        try:
+            cythonize(
+                self.distribution.cython_manual_modules,
+                replace_extension=True
```
I've fixed it and added a test

scoder commented:
This feature depends on configuration from the installed Cython package, right? We're not currently installing Cython for the test runs, so the tests will probably fail like this. I think we need a virtualenv here. That would also make sure `setuptools` is available.

For simplicity, I think it's ok if the tests only run in Py3, so you can use `PYTHON -m venv …` and then continue from there, `venv/bin/python3 setup.py install` etc. It's at least pre-compiled, so this should be somewhat fast. I just added a tag `py3only` in master, so you can rebase (again, sorry) and then tag your tests as  `# tag: py3only` at the top to prevent Py2 from picking them up.

nils-werner commented:
> We're not currently installing Cython for the test runs, so the tests will probably fail like this.

I don't understand, how can you even test a library that's not installed?

scoder commented:
> how can you even test a library that's not installed?

It's there, in the checkout. The tests run from the project base directory.

nils-werner commented:
Right. I am just wondering because I thought it was a common pattern to

    pip install -e .

any library before running its tests.

scoder commented:
> `pip install -e .`

Then you still need to make sure that the project source directory is not on the `PYTHONPATH`, which usually involves switching to a different directory or moving all sources into a subdirectory. None of these is currently done in Cython.

However, this problem doesn't apply here since the installation creates new files, so there is no overlap with existing sources.

nils-werner commented:
Are you sure? I may be mistaken, but to my knowledge `pip install -e .` does not create new files, but instead just compiles any extensions and adds the local library path to `PYTHONPATH`. Afterwards it can be imported from anywhere, including inside the base checkout.

scoder commented:
Ah, sorry, right. It installs an egg-link file that points to the current directly. It still needs to put the endpoint scripts somewhere, though. That's what I meant with "generate files".

nils-werner commented:
Thanks for approving the code. Regarding docs:

Should the docs only mention this method as an alternative, or should it (at some point in the future) be mentioned as the main method?

auvipy commented:
I think initially alternative

scoder commented:
I would still like to see a venv being used in the tests in order to test that as well. Also, running a test should not depend on running a pip-dev install first.

scoder commented:
Please prefix the test names with `setuptools_…`. `cython_modules` is not a very specific name in the context of Cython test modules.

scoder commented:
Regarding docs, I think a short examle can go into `quickstart/build.rst`, in addition (and right after) the current way. And there is more than one place in `userguide/source_files_and_compilation.rst` where it's worth mentioning this feature, with a main section before "Configuring the C build".

nils-werner commented:
Thanks for the feedback! I have added some sections to the docs you've mentioned, and renamed the tests.

> I would still like to see a venv being used in the tests in order to test that as well.

Is this related to this PR, or a thing that should be tested in any case?

> Also, running a test should not depend on running a pip-dev install first.

For these tests I see no way around it, as this is a feature of a pip install. For all other tests I don't see a problem as it doesn't seem to make a difference.

scoder commented:
> Is this related to this PR, or a thing that should be tested in any case?

Both. It's specifically related to this PR, because, as you wrote:

> For these tests I see no way around it, as this is a feature of a pip install.

nils-werner commented:
I just checked and turns out Travis already runs in virtualenvs by default: i.e. [Python 3.8](https://travis-ci.org/github/cython/cython/jobs/684625897#L355) and [Python 2.7](https://travis-ci.org/github/cython/cython/jobs/684625899#L352)

MuellerSeb commented:
Is this still needed? We now got `pyproject.toml` to specify build time dependencies, so Cython should be available when setup.py is executed.

scoder commented:
Related to #2531
